### DISCUSSION TITLE
Enabling DAG Runs and Task Instance permissions to a Custom Role via code

### DISCUSSION THREAD

**Discussion by dijana-z** [apache/airflow#32236] [Q&A] [answered]
Hello,

I am using Apache Airflow version 2.2.4 and am trying to modify `access_control` for DAGs so that the custom role I have created has permissions to edit and view DAG Runs and Task Instances. 

Previously, I've added DAG Level permissions this way:
```
DAG(
...
access_control={"custom_role" : {"can_edit", "can_read"}}
)
```

Since I have custom AD Groups that correspond to the custom roles I have, I want only those users to have permissions for certain DAGs and therefore to DAG Runs and Task Instances.
When I tried to add permissions like this (or similar ways):
```
DAG(
...
access_control = {"custom_role" : {
            (permissions.ACTION_CAN_READ, permissions.RESOURCE_DAG),
            (permissions.ACTION_CAN_EDIT, permissions.RESOURCE_DAG),
            (permissions.ACTION_CAN_READ, permissions.RESOURCE_DAG_RUN ),
            (permissions.ACTION_CAN_READ, permissions.RESOURCE_TASK_INSTANCE),
            (permissions.ACTION_CAN_EDIT, permissions.RESOURCE_DAG_RUN ),
            (permissions.ACTION_CAN_EDIT, permissions.RESOURCE_TASK_INSTANCE),
        }
    }
)
```
I get an error that was raised [here](https://github.com/apache/airflow/blob/2.2.4/airflow/www/security.py#L840).

As I saw, the same check and error is still being raised on newer Airflow versions.

All examples I saw always deal with DAG Level permissions, so my question is - is there a way to modify more granular permissions for custom roles, and if so - how? 

Thank you!

**Comment by potiuk** [comment, marked as answer]
No. Not now. But we just approved AIP-56 "Extensible user management"  https://cwiki.apache.org/confluence/display/AIRFLOW/AIP-56+Extensible+user+management and the way it is designed it will allow those who will manage the deployments to integrate their authentication / authorization in a much more flexible way  - utilsing integration via external authentication solutions such as KeyCloak. This is a strategic direction we are taking, and once implemented in future versions of Airlfow you will be able to do similar mapping of user roles to capabilities for certain groups for example.

**Reply by dijana-z** [reply]
Thanks for the info!

**Reply by RenGeng** [reply]
Hello @potiuk, do you know when this feature will be added ?

**Reply by potiuk** [reply]
First iteration likely to be present in Airlfow 2.8 - but generally the answer is "when ready".

**Reply by potiuk** [reply]
Like everything in OSS - things are released when ready, not when there is a roadmap-that-nobody-pays-for-to-have-and-keep.

**Reply by hakan-77** [reply]
@potiuk, thank you all for the amazing new security framework. 

Is AIP-56 work completed or still work in progress?

I am using 2.8 and curious if it is possible to control DAG access per user. 

FabAirflowSecurityManagerOverride has methods such as get_readable_dags() but somehow overriding these methods are not working. They are never called.

**Reply by potiuk** [reply]
I think you can try to do it, but you are on your own. The API is not closed yet and people are still working on it. documentation is in progress, separation of FAB to a separate provider, and a lot more testing is needed and writing an alternative security backend to prove it works. So until it's all settled, you are mostly on your own experimenting with it (and hopefully also contributing fixes if you find and fix them).
